### PR TITLE
User/aeloros/crashandforeground

### PR DESCRIPTION
**Pull request by aeloros**
Ensure all static calls initialize the singleton for the current instance.

Attempt to transfer foreground rights when  redirecting an activation to another instance.

Updated a useful manual test app to have UI.

### CODE REVIEW

File: test/TestApps/ManualTestApp/main.cpp [hunk 1]
```diff
@@ -36,10 +56,10 @@ HRESULT BootstrapInitialize()
...
     RETURN_IF_FAILED(MddBootstrapTestInitialize(c_PackageNamePrefix, c_PackagePublisherId));
 
-    // Version <major>.0.0.0 to find any framework package for this major version
-    const UINT64 c_Version_Major{ 4 };
-    PACKAGE_VERSION minVersion{ static_cast<UINT64>(c_Version_Major) << 48 };
-    RETURN_IF_FAILED(MddBootstrapInitialize(c_Version_Major, nullptr, minVersion));
+    // Major.Minor version, MinVersion=0 to find any framework package for this major.minor version
+    const UINT32 c_Version_MajorMinor{ 0x00040001 };
+    const PACKAGE_VERSION minVersion{};
+    RETURN_IF_FAILED(MddBootstrapInitialize(c_Version_MajorMinor, nullptr, minVersion));
```
Comment (jonwis) [inline test/TestApps/ManualTestApp/main.cpp:62]: @DrusTheAxe - when you go add your generated versioning headers, please make a pass and find and fix anybody calling Mdd with manual parameters in samples & tests.

**Review by jonwis — APPROVED** [review]

File: test/TestApps/ManualTestApp/main.cpp [hunk 2]
```diff
@@ -14,15 +14,35 @@ using namespace winrt::Microsoft::Windows::AppLifecycle;
...
 
-bool IsPackagedProcess()
+HWND g_window = NULL;
+wchar_t g_windowClass[] = L"TestWndClass"; // the main window class name
+
+ATOM _RegisterClass(HINSTANCE hInstance);
+BOOL InitInstance(HINSTANCE, int);
+LRESULT CALLBACK WndProc(HWND, UINT, WPARAM, LPARAM);
+
+std::wstring GetFullIdentityString()
```
Comment (DrusTheAxe) [inline test/TestApps/ManualTestApp/main.cpp:24]: 'GetFullIdentityString' is `GetCurrentPackageFullName but return as a std::wstring`?

Even better if added to `dev\common\AppModel.Identity.h`

```
namespace appModel::Identity
{
inline bool IsPackagedProcess()
{
    ...
}

inline std::wstring GetCurrentPackageFullName()
{
    WCHAR packageFullName[PACKAGE_FULL_NAME_MAX_LENGTH + 1]{};
    UINT32 n{ static_cast<UINT32>(ARRAYSIZE(packageFullName)) };
    const auto rc{ ::GetCurrentPackageFullName(&n, packageFullName) };
    if (rc == APPMODEL_ERROR_NO_PACKAGE)
    {
        return std::wstring();
    }
    THROW_IF_WIN32_ERROR(rc);
    return std::wstring(packageFullName);
}
...
```

File: test/TestApps/ManualTestApp/main.cpp [hunk 3]
```diff
@@ -14,15 +14,35 @@ using namespace winrt::Microsoft::Windows::AppLifecycle;
...
+    UINT32 idNameBufferLen = ARRAYSIZE(idNameBuffer);
+    if (::GetCurrentPackageFullName(&idNameBufferLen, idNameBuffer) == ERROR_SUCCESS)
+    {
+        identityString = idNameBuffer;
+    }
+
+    return identityString;
+}
+
+bool HasIdentity()
```
Comment (DrusTheAxe) [inline test/TestApps/ManualTestApp/main.cpp:37]: Remove and changed callers to

```
#include <appmodel.identity.h>
...
if (AppModel::Identity::IsPackagedProcess())
```

File: test/TestApps/ManualTestApp/main.cpp [hunk 4]
```diff
@@ -72,34 +93,98 @@ int main()
...
     {
-        if (myKeyInstance.IsCurrent())
+        AppInstance thisInstance = AppInstance::GetCurrent();
+        token = thisInstance.Activated(
+            auto_revoke, [&thisInstance](
+                const auto& sender, const AppActivationArguments& args)
+            { OnActivated(sender, args); }
+        );
+
+        auto hInstance = GetModuleHandle(NULL);
```
Comment (DrusTheAxe) [inline test/TestApps/ManualTestApp/main.cpp:112]: NIT: `nullptr`

**Review by DrusTheAxe — APPROVED** [review]
Some refactoring of the identity functions but otherwise LGTM

File: dev/AppLifecycle/AppInstance.cpp [hunk 5]
```diff
@@ -216,8 +216,11 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
         wil::unique_event cleanupEvent;
         cleanupEvent.create(wil::EventOptions::ManualReset, eventName.c_str());
 
-        // Enqueue the request and signal the other instance.
+        // Enqueue the request and transfer foreground rights.
         EnqueueRedirectionRequestId(id);
+        AllowSetForegroundWindow(m_processId);
```
Comment (EHO-makai) [inline dev/AppLifecycle/AppInstance.cpp:221]: I assume you don't care if this succeeds or not? (ignoring BOOl return value)

**Review by EHO-makai — APPROVED** [review]

**Pull request merged**